common/txmgr/confirmer.go

```diff
@@ -294,178 +260,151 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
+		} else {
+			ec.lggr.Infow(fmt.Sprintf("Re-org detected. Rebroadcasting transaction %s which may have been re-org'd out of the main chain", attempt.Hash.String()), logValues...)
 		}
 
-		// separateLikelyConfirmedAttempts is used as an optimisation: there is
-		// no point trying to fetch receipts for attempts with a sequence higher
-		// than the highest sequence the RPC node thinks it has seen
-		likelyConfirmed := ec.separateLikelyConfirmedAttempts(from, attempts, minedSequence)
-		likelyConfirmedCount := len(likelyConfirmed)
-		if likelyConfirmedCount > 0 {
-			likelyUnconfirmedCount := len(attempts) - likelyConfirmedCount
+		etxIDs = append(etxIDs, etx.ID)
+		attemptIDs = append(attemptIDs, attempt.ID)
+	}
 
-			ec.lggr.Debugf("Fetching and saving %v likely confirmed receipts. Skipping checking the others (%v)",
-				likelyConfirmedCount, likelyUnconfirmedCount)
+	// Mark transactions as unconfirmed, mark attempts as in-progress, and delete receipts since they do not apply to the new chain
+	// This may revert some fatal error transactions to unconfirmed if terminally stuck transactions purge attempts get re-org'd
+	return ec.txStore.UpdateTxForRebroadcast(ctx, etxIDs, attemptIDs)
+}
 
-			start := time.Now()
-			err = ec.fetchAndSaveReceipts(ctx, likelyConfirmed, blockNum)
-			if err != nil {
-				return fmt.Errorf("unable to fetch and save receipts for likely confirmed txs, for address: %v: %w", from, err)
+func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) ProcessIncludedTxs(ctx context.Context, includedTxs []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], head types.Head[BLOCK_HASH]) error {
+	if len(includedTxs) == 0 {
+		return nil
+	}
+	// Add newly confirmed transactions to the prom metric
+	promNumConfirmedTxs.WithLabelValues(ec.chainID.String()).Add(float64(len(includedTxs)))
+
+	var purgeTxIDs []int64
+	var confirmedTxIDs []int64
+	for _, tx := range includedTxs {
+		isPurgeTx := false
+		for _, attempt := range tx.TxAttempts {
+			if attempt.IsPurgeAttempt {
+				isPurgeTx = true
+				break
 			}
-			ec.lggr.Debugw(fmt.Sprintf("Fetching and saving %v likely confirmed receipts done", likelyConfirmedCount),
-				"time", time.Since(start))
 		}
+		// If any attempt in the transaction is marked for purge, the transaction was terminally stuck and should be marked as fatal error
+		if isPurgeTx {
+			purgeTxIDs = append(purgeTxIDs, tx.ID)
+			continue
+		}
+		confirmedTxIDs = append(confirmedTxIDs, tx.ID)
+		observeUntilTxConfirmed(ec.chainID, tx.TxAttempts, head)
 	}
-
-	if err := ec.txStore.MarkAllConfirmedMissingReceipt(ctx, ec.chainID); err != nil {
-		return fmt.Errorf("unable to mark txes as 'confirmed_missing_receipt': %w", err)
+	// Mark the transactions included on-chain with a purge attempt as fatal error with the terminally stuck error message
+	if err := ec.txStore.UpdateTxFatalError(ctx, purgeTxIDs, ec.stuckTxDetector.StuckTxFatalError()); err != nil {
```

Comment:
I'm wondering if this functionality is better suited for the Finalizer instead

Reply:
I think we would want to mark tx as fatal for being stuck as soon as possible which is why this was left in the Confirmer. Otherwise, we would only surface the fatal error after finalization which could take a long time. I also wanted to avoid marking a fatal stuck transaction as confirmed at any point to avoid any confusion with what that state implies.